### PR TITLE
Fire an event when allowing a bad request 

### LINKED ISSUE #39756: Diagnostics for EnableInsecureAbsoluteFormHostOverride
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

https://github.com/dotnet/aspnetcore/pull/39334 added a 6.0 servicing workaround for clients that sent a certain kind of invalid requests. The affected client is being updated (slowly) so we were not planning to port the fix to 7.

The customer has asked for some additional diagnostics so they can know if anybody else is hitting this issue (without disabling the mitigation), and that would help inform if we need to port the fix to 7.0.

### Describe the solution you'd like

Add some diagnostics (logs or events) for when this mitigation is hit so we can track how many & who would be affected if it were disabled. We need to work with the customer to find the most consumable output.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Tratcher**
# Fire an event when allowing a bad request 

This adds diagnostics for a prior fix so partners can determine when the client issues have been addressed.

## Description

https://github.com/dotnet/aspnetcore/pull/39334 from Feb allowed a partner to opt into accepting a specific kind of bad request. They're working to address the issues with the remote client and aren't sure how long that will take. They've requested new event so they can add in their own logging to monitor the client's progress, detect if anyone else is encountering the issue, and decide if we need a long term fix added to 7.0.

Fixes #39756

## Customer Impact

The customer is unable to tell if the prior mitigation is still required without disabling it and potentially breaking live customers.

## Regression?

- [ ] Yes
- [x] No

## Risk

- [ ] High
- [ ] Medium
- [x] Low

[Justify the selection above]

## Verification

- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 1]
```diff
@@ -1340,6 +1340,21 @@ public void SetBadRequestState(BadHttpRequestException ex)
             _keepAlive = false;
         }
 
+        // Normally this would have been rejected, but the developer opted into allowing the bad behavior.
+        public void ReportAllowedBadRequest(BadHttpRequestException ex)
+        {
+            Log.ConnectionBadRequest(ConnectionId, ex);
+            _requestRejectedException = ex;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1347]:
```suggestion
            // Set this so DiagnosticSource listeners can observe the exception via the IBadRequestExceptionFeature.
            // Make sure to unset before exiting the method so the request doesn't actually get rejected.
            // This means the exception will not be visible to normal middleware.
            _requestRejectedException = ex;
```

Alternatively, we could create another field for `_allowedBadRequestException` or something so it's visible from middleware too. I don't think that's necessary though.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 2]
```diff
@@ -1340,6 +1340,24 @@ public void SetBadRequestState(BadHttpRequestException ex)
...
+            // This means the exception will not be visible to normal middleware.
+            _requestRejectedException = ex;
+
+            const string badRequestEventName = "Microsoft.AspNetCore.Server.Kestrel.BadRequest";
+            if (ServiceContext.DiagnosticSource?.IsEnabled(badRequestEventName) == true)
+            {
+                ServiceContext.DiagnosticSource.Write(badRequestEventName, this);
+            }
+
+            _requestRejectedException = null;
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1358]: What happens if an exception is thrown from DiagnosticSource.Write? Do we care that _requestRejectedException isn't null?

tldr should it be a try/finally?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1358]: I think that exception would cause the request to fail anyways, so _requestRejectedException wouldn't matter.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 3]
```diff
@@ -1340,6 +1340,24 @@ public void SetBadRequestState(BadHttpRequestException ex)
             _keepAlive = false;
         }
 
+        // Normally this would have been rejected, but the developer opted into allowing the bad behavior.
+        public void ReportAllowedBadRequest(BadHttpRequestException ex)
+        {
+            Log.ConnectionBadRequest(ConnectionId, ex);
```
Comment (danmoseley) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1346]: this is new logging that now occurs even without setting the appcontext switch?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1346]: This method is only called when the appcontext switch is set. Compare to the SetBadRequestState method above that's called when the appcontext switch is not set. Both write the same log, but the new one avoids marking the request as failed/rejected.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:1358]: There are different paths that a request goes down on failure when this property is sent. e.g. writing a 400 response
